--- htr.c
+++ search.c
@@ -10,9 +10,10 @@
   return ~ret + (uint64_t) 1;
 }
 // search hard-to-round cases of exp in [x0,x1)
-// with at least m0 identical bits after round bit
-static void
-search (double x0, double x1, int m)
+// with at least m identical bits after round bit
+void
+search (eval_t eval, expo_t expo, report_t report,
+        double x0, double x1, int m)
 {
   uint64_t n = 1ul << 20;
   int e, e0, e1;
@@ -25,17 +26,14 @@
   double ux = ldexp (1.0, e0 - 53); // ux = ulp(x)
 
   // check exp(x) lies in the same binade in [x0, x1)
-  double y0 = ref_exp (x0);
-  double y1 = ref_exp (nextafter (x1, x0));
-  frexp (y0, &e);
-  frexp (y1, &e1);
+  e = expo (x0);
+  e1 = expo (nextafter (x1, x0));
   assert (e == e1);
-  double uh = ldexp (1.0, e - 53); // uh = ulp(exp(x))
 
   double h, l, s;
   double x = x0;
   while (x < x1) {
-    dd_exp (&h, &l, &s, x);
+    eval (&h, &l, &s, x);
     double dh = h * ux, dl = l * ux; // 1st derivative, multiplied by ux
     double dd = h * ux * ux; // 2nd derivative, multiplied by ux^2
 
@@ -65,7 +63,7 @@
     for (uint64_t i = 0; i < n; i++) {
       if (__builtin_expect (A < 2*E, 0)) { // found potential hard-to-round case
         double xi = x + i * ux;
-        check (xi, m);
+        report (xi, m);
       }
       A += B;
     }
